File: RobWorkSim/src/sandbox/grippers/GraspPlugin/createGripperDB.cpp
#include <iostream>
#include <rw/rw.hpp>
#include <rw/loaders/model3d/STLFile.hpp>
#include <boost/program_options/options_description.hpp>
#include <boost/program_options/variables_map.hpp>
#include <boost/program_options/option.hpp>
#include <boost/program_options/parsers.hpp>
#include "Gripper.hpp"
#include "GripperXMLLoader.hpp"
#include "JawPrimitive.hpp"



using namespace std;
USE_ROBWORK_NAMESPACE;
using namespace robwork;
using namespace rw::models;
using namespace rw::pathplanning;
using namespace boost::program_options;
namespace po = boost::program_options;



int main(int argc, char* argv[])
{
	// parameters
	int number;
	Q low(10, 	0,	0.01, 	0.001, 	0.001, 	0, 	0, 	0, 		0, 		60, 	0);
	Q high(10, 	1,	0.2, 	0.05, 	0.05, 	1, 	90, 1, 		0.05, 	120, 	0.25);
	
	// program options
	options_description desc("Allowed options");
	desc.add_options()
		("help,h", "help message")
		("number,n", value<int>(&number)->default_value(1), "number of grippers to generate [1]")
		("base,b", value<vector<double> >()->multitoken(), "bases description; 3 parameters should follow")
		("length,l", value<vector<double> >()->multitoken(), "min & max length [0.01; 0.2]")
		("width,w", value<vector<double> >()->multitoken(), "min & max width [0.001; 0.05]")
		("depth,d", value<vector<double> >()->multitoken(), "min & max depth [0.001; 0.05]")
		("chfdepth", value<vector<double> >()->multitoken(), "min & max chamfer depth <0; 1>")
		("chfangle", value<vector<double> >()->multitoken(), "min & max chamfer angle <0; 90>")
		("cutdepth", value<vector<double> >()->multitoken(), "min & max cut depth")
		("cutangle", value<vector<double> >()->multitoken(), "min & max cut angle <0; 180>")
		("cutradius", value<vector<double> >()->multitoken(), "min & max cut radius [0; 0.25]")
		("tcp,t", value<vector<double> >()->multitoken(), "min & max tcp position")
		("opening,o", value<vector<double> >()->multitoken(), "min & max opening")
		("force,o", value<vector<double> >()->multitoken(), "min & max opening")
		("stl", "save base & jaw meshes to STL")
	;
	positional_options_description posDesc;
	posDesc.add("out", -1);
	
	try {
		variables_map vm;
		store(command_line_parser(argc, argv).options(desc).positional(posDesc).run(), vm);
		notify(vm);
		
		/* PROCESS */
		if (vm.count("help")) {
			cout  	<< "This is a script creating gripper XML files database.\n"
					<< "A number of grippers is randomly generated within given constraints.\n" << endl;

			cout << desc << endl;
			return 1;
		}
		
		// setup sampler
		Device::QBox box(low, high);
		QSampler::Ptr jawSampler = QSampler::makeUniform(box);
		
		for (int i = 0; i < number; ++i) {
			Gripper::Ptr gripper = ownedPtr(new Gripper);
			
			stringstream sname;
			sname << "gripper" << i;
			gripper->setName(sname.str());
			
			Q jawParams = jawSampler->sample();
			jawParams(0) = int(jawParams(0));
			jawParams(6) = jawParams(6) * jawParams(1);
			cout << jawParams << endl;
			gripper->setJawGeometry(jawParams);
			
			GripperXMLLoader::save(gripper, ".", gripper->getName()+".grp.xml");
			if (vm.count("stl")) {
				STLFile::save(*gripper->getJawGeometry()->getGeometryData()->getTriMesh(), gripper->getName()+"_jaw.stl");
				STLFile::save(*gripper->getBaseGeometry()->getGeometryData()->getTriMesh(), gripper->getName()+"_base.stl");
			}
		}
		
	} catch (...) {
		cout << "ERROR! Usage:\n" << endl;
		cout << desc << endl;
		return -1;
	}
	
	return 0;
}

File: RobWorkSim/src/sandbox/grippers/GraspPlugin/bin/create-gripper-db.py
#!/usr/bin/python

import sys
import os
import math
import random
import argparse



class GripperGenerator:
	"Class for generating grippers."
	
	def __init__(self):
		self.gripperName = "gripper"
		self._gripperId = 0
		self.length = (0.01, 0.2)
		self.width = (0.001, 0.05)
		self.depth = (0.001, 0.05)
		self.chfChance = 0.5
		self.chfDepth = (0, 1)
		self.chfAngle = (15, 75)
		self.tcpOffset = (0.2, 0.8)
		self.cutChance = 0.67
		self.cutDepth = (0, 0.05)
		self.cutAngle = (60, 120)
		self.cutRadius = (0, 0.1)
		self.opening = (0.05, 0.1)
		self.jawdist = (0, 0.01)
		self.force = (25, 75)
		
	def generate(self, stl=False):
		# setup gripper name
		name = self.gripperName + str(self._gripperId)
		self._gripperId += 1
		
		# randomize general dimensions
		cutout = random.randint(0, 1)
		length = random.uniform(*self.length)
		width = random.uniform(*self.width)
		depth = random.uniform(*self.depth)
		
		chfDepth = 0
		chfAngle = 0
		if random.uniform(0, 1) < self.chfChance:
			chfDepth = random.uniform(*self.chfDepth)
			chfAngle = random.uniform(*self.chfAngle)
			
		tcpOffset = random.uniform(*self.tcpOffset)
		cutAngle = random.uniform(*self.cutAngle)
		opening = random.uniform(*self.opening)
		force = random.uniform(*self.force)
		
		# find tcp position
		tcp = tcpOffset * length
		
		# generate cut depth
		cutDepth = 0
		if random.uniform(0, 1) < self.cutChance:
			d = length - chfDepth * width * math.tan(math.radians(chfAngle))
			lwidth = width
			if tcp > d:
				lwidth = width - (tcp - d) * 1.0/math.tan(math.radians(chfAngle))
			cutDepth = random.uniform(self.cutDepth[0], min(self.cutDepth[1], lwidth-0.001))
		
		# generate cut radius
		cutRadius = random.uniform(max(self.cutRadius[0], cutDepth), self.cutRadius[1])
		
		# generate jawdist
		jawdist = random.uniform(self.jawdist[0], min(self.jawdist[1], opening))
		
		# make jaw parameters
		jawParameters = (cutout, length, width, depth, chfDepth, chfAngle, tcp, cutDepth, cutAngle, cutRadius)
		jawStr = "".join(str(p)+' ' for p in jawParameters)
		
		# construct command
		cd = "${RWSIM_ROOT}src/sandbox/grippers/GraspPlugin/bin/"
		cmd = cd+"create-gripper-xml --name {} -j {} --tcp {} --jawdist {} --opening {} --force {} {}"\
			.format(name, jawStr, tcp, jawdist, opening, force, ("", "--stl")[stl])
			
		# execute
		#print cmd
		os.system(cmd)

def main():
	parser = argparse.ArgumentParser(description="Generates gripper database.")
	parser.add_argument("number", nargs="?", default="1", help="number of grippers to generate")
	parser.add_argument("--stl", action="store_true", help="generate STL files")
	args = parser.parse_args()
	
	generator = GripperGenerator()
	
	loading = ['|', '/', '-', '\\']
	sys.stdout.write("Generating gripper database...  ")
	
	for i in range(0, int(args.number)):
		generator.generate(bool(args.stl))
		sys.stdout.write("\b"+loading[i%4])
		sys.stdout.flush()

	sys.stdout.write("\b Done.\n")

if __name__ == "__main__":
	main()
